### PR TITLE
WWZ4l Run3

### PR DESCRIPTION
**Pull request by MatthewDittrich**
Run3 (2022) additions for the wwz4l processor

### CODE REVIEW

File: ewkcoffea/modules/objects_wwz.py [hunk 1]
```diff
@@ -17,32 +17,40 @@ def get_cleaned_collection(obj_collection_a,obj_collection_b,drcut=0.4):
...
+        (ele.miniPFRelIso_all < get_ec_param("wwz_pres_e_miniPFRelIso_all")) &
         (abs(ele.sip3d)       <  get_ec_param("wwz_pres_e_sip3d")) &
-        (ele.miniPFRelIso_all <  get_ec_param("wwz_pres_e_miniPFRelIso_all")) &
         (ele.lostHits         <= get_ec_param("wwz_pres_e_lostHits"))
     )
-    if tight: mask = (mask & ele.convVeto & (ele.tightCharge == get_ec_param("wwz_pres_e_tightCharge")))
-    return mask
+    if ("2022" in year): mask_return = (mask & ele.mvaIso_WP80)
+    if ("2022" not in year): mask_return = (mask & mask_run2)
+    if tight: mask_return = (mask_return & ele.convVeto & (ele.tightCharge == get_ec_param("wwz_pres_e_tightCharge")))
```
Comment (kmohrman) [inline ewkcoffea/modules/objects_wwz.py:34]: Is this "tight" mask is relevant and is wanted to be applied to R3?

File: ewkcoffea/modules/objects_wwz.py [hunk 2]
```diff
@@ -17,32 +17,40 @@ def get_cleaned_collection(obj_collection_a,obj_collection_b,drcut=0.4):
...
+    )
+    mask_run2 = (
+        (ele.miniPFRelIso_all < get_ec_param("wwz_pres_e_miniPFRelIso_all")) &
         (abs(ele.sip3d)       <  get_ec_param("wwz_pres_e_sip3d")) &
-        (ele.miniPFRelIso_all <  get_ec_param("wwz_pres_e_miniPFRelIso_all")) &
         (ele.lostHits         <= get_ec_param("wwz_pres_e_lostHits"))
     )
-    if tight: mask = (mask & ele.convVeto & (ele.tightCharge == get_ec_param("wwz_pres_e_tightCharge")))
-    return mask
+    if ("2022" in year): mask_return = (mask & ele.mvaIso_WP80)
```
Comment (kmohrman) [inline ewkcoffea/modules/objects_wwz.py:32]: I think this would be more clear with an if an and else, along with the statements being on their own lines. I.e.:
```
if (2022 in year):
    stuff
else:
    other stuff

File: ewkcoffea/modules/objects_wwz.py [hunk 3]
```diff
@@ -17,32 +17,40 @@ def get_cleaned_collection(obj_collection_a,obj_collection_b,drcut=0.4):
...
-        (abs(mu.sip3d)       <  get_ec_param("wwz_pres_m_sip3d")) &
-        (mu.miniPFRelIso_all <  get_ec_param("wwz_pres_m_miniPFRelIso_all")) &
         (mu.mediumId)
     )
-    return mask
+    mask_run2 = (
+        (mu.miniPFRelIso_all < get_ec_param("wwz_pres_m_miniPFRelIso_all")) &
+        (abs(mu.sip3d)       <  get_ec_param("wwz_pres_m_sip3d"))
+    )
+    if ("2022" not in year): mask_return = (mask & mask_run2)
```
Comment (kmohrman) [inline ewkcoffea/modules/objects_wwz.py:51]: Same comment as for ele

File: ewkcoffea/modules/selection_wwz.py [hunk 4]
```diff
@@ -68,6 +68,7 @@
             "Ele32_WPTight_Gsf_L1DoubleEG",
             "Ele35_WPTight_Gsf",
             "Ele115_CaloIdVT_GsfTrkIdT",
+            "DoublePhoton70",
```
Comment (kmohrman) [inline ewkcoffea/modules/selection_wwz.py:71]: Was this intended to be added? Does it make any impact?

File: ewkcoffea/modules/selection_wwz.py [hunk 5]
```diff
@@ -175,6 +176,39 @@
         "EGamma"         : dataset_dict["2018"]["DoubleMuon"],
         "MuonEG"         : dataset_dict["2018"]["DoubleMuon"] + dataset_dict["2018"]["EGamma"],
     },
+    "B": {
```
Comment (kmohrman) [inline ewkcoffea/modules/selection_wwz.py:179]: If we are going to have the trigger list and exclude dict here, it should be removed from `selection_run3_2lep.py` (we definitely do not want to store it in two places). The other option is to add it as a json that both can reference. However, jsons do not allow comments and some of the things here we really want comments for (e.g. the Era C complications).

So I would suggest leaving the dictionary here (and then just removing it from the dilepton section file). The di lepton processor can just point to this copy here.

File: ewkcoffea/modules/selection_wwz.py [hunk 6]
```diff
@@ -227,6 +260,7 @@ def add4lmask_wwz(events, year, isData, sample_name):
 
     # Filters
     filter_flags = events.Flag
+    # TODO: Should we have a different list of filters for Run3? Filter effect seems to be quite small
```
Comment (kmohrman) [inline ewkcoffea/modules/selection_wwz.py:263]: I would say we should go with the ones Guillelmo applies. Do you have that information?

File: analysis/wwz/wwz4l.py [hunk 7]
```diff
@@ -206,9 +208,13 @@ def process(self, events):
             sow_factDown       = -1
 
 
-        datasets = ["SingleMuon", "SingleElectron", "EGamma", "MuonEG", "DoubleMuon", "DoubleElectron", "DoubleEG"]
+        datasets = ["SingleMuon", "SingleElectron", "EGamma", "MuonEG", "DoubleMuon", "DoubleElectron", "DoubleEG","Muon"]
         for d in datasets:
-            if d in dataset: dataset = dataset.split('_')[0]
+            if d in dataset:
```
Comment (kmohrman) [inline analysis/wwz/wwz4l.py:213]: Could you explain what this is doing?

File: analysis/wwz/wwz4l.py [hunk 8]
```diff
@@ -217,6 +223,8 @@ def process(self, events):
         mu   = events.Muon
         tau  = events.Tau
         jets = events.Jet
+        if ("2022" in year) and (not isData):
```
Comment (kmohrman) [inline analysis/wwz/wwz4l.py:226]: But events.Pilepu is used for non 2022 years too, right? Is there a particular reason this if is only applied for 2022?

Also, it looks like `pileup` is only used in a few limited places. Would it be easier to just use `events.Pileup` there, as we do not R2?

File: analysis/wwz/wwz4l.py [hunk 9]
```diff
@@ -319,10 +338,13 @@ def process(self, events):
...
             weights_obj_base.add('fact', events.nom, events.factUp*(sow/sow_factUp), events.factDown*(sow/sow_factDown))
-
-            # Misc other experimental SFs and systs
-            weights_obj_base.add('PreFiring', events.L1PreFiringWeight.Nom,  events.L1PreFiringWeight.Up,  events.L1PreFiringWeight.Dn)
-            weights_obj_base.add('PU', cor_tc.GetPUSF((events.Pileup.nTrueInt), year), cor_tc.GetPUSF(events.Pileup.nTrueInt, year, 'up'), cor_tc.GetPUSF(events.Pileup.nTrueInt, year, 'down'))
+            if "2022" not in year:
+                # Misc other experimental SFs and systs
+                weights_obj_base.add('PreFiring', events.L1PreFiringWeight.Nom,  events.L1PreFiringWeight.Up,  events.L1PreFiringWeight.Dn)
+                weights_obj_base.add('PU', cor_tc.GetPUSF((events.Pileup.nTrueInt), year), cor_tc.GetPUSF(events.Pileup.nTrueInt, year, 'up'), cor_tc.GetPUSF(events.Pileup.nTrueInt, year, 'down'))
+            if "2022" in year:
```
Comment (kmohrman) [inline analysis/wwz/wwz4l.py:345]: I think this would make more sense (and follow your previous if statements more closely) if this were an `else`. Or is there some reason I'm missing for why you are not using an else here?

File: analysis/wwz/wwz4l.py [hunk 10]
```diff
@@ -361,6 +385,7 @@ def process(self, events):
 
             # Selecting jets and cleaning them
             jetptname = "pt_nom" if hasattr(cleanedJets, "pt_nom") else "pt"
+            # TODO: Check there are no Run3 issues with this
```
Comment (kmohrman) [inline analysis/wwz/wwz4l.py:388]: What does this refer to?

File: analysis/wwz/wwz4l.py [hunk 11]
```diff
@@ -404,7 +435,7 @@ def process(self, events):
 
             ######### Apply SFs #########
 
-            if not isData:
+            if (not isData):
```
Comment (kmohrman) [inline analysis/wwz/wwz4l.py:438]: I don't think this is necessary

**Comment by kmohrman** [comment]
As a general comment on the processor, I think we should replace all instances of "`if "2022" in year` with a variable. I.e. towards the beginning of the processor, do this:
```
# Define a flag to use if this is a 2022 sample
if year in ["2022", "2022EE"]:
    is2022 = True
else:
    is2022 = False
```
Then throughout the processor can use `if is2022. This is more explicit and I think more readable than the current method.

File: analysis/wwz/wwz4l.py [hunk 12]
```diff
@@ -745,71 +784,82 @@ def process(self, events):
...
             # List the hists that are only defined for some categories
-            analysis_cats = ["sr_4l_sf_A","sr_4l_sf_B","sr_4l_sf_C","sr_4l_of_1","sr_4l_of_2","sr_4l_of_3","sr_4l_of_4"] + bdt_sr_names
+            analysis_cats = ["sr_4l_sf_A","sr_4l_sf_B","sr_4l_sf_C","sr_4l_of_1","sr_4l_of_2","sr_4l_of_3","sr_4l_of_4"]
+            if "2022" not in year:
+                analysis_cats = analysis_cats + bdt_sr_names
             exclude_var_dict = {
                 "mt2" : ["all_events"],
                 "ptl4" : ["all_events"],
-                "j0pt" : ["all_events", "4l_presel", "sr_4l_sf_incl", "sr_4l_of_incl", "sr_4l_bdt_sf_presel", "sr_4l_bdt_sf_trn", "sr_4l_bdt_of_trn", "sr_4l_bdt_of_presel", "cr_4l_sf"] + analysis_cats,
+                "j0pt" : ["all_events", "4l_presel", "sr_4l_sf_incl", "sr_4l_of_incl", "cr_4l_sf"] + analysis_cats,
```
Comment (kmohrman) [inline analysis/wwz/wwz4l.py:862]: We still need to exclude the SR categories here, right?

**Comment by kmohrman** [comment]
Have you added config files that point to all of the R3 jsons? If that's not in yet, can you ping me when it's ready so I can try to run on your branch?

File: ewkcoffea/modules/objects_wwz.py (hunk 1 above)
Comment (MatthewDittrich) [reply ewkcoffea/modules/objects_wwz.py:34]: I can remove this as a v0 for Run3

File: ewkcoffea/modules/selection_wwz.py (hunk 4 above)
Comment (MatthewDittrich) [reply ewkcoffea/modules/selection_wwz.py:71]: Technically, Guillermo is using this. I can perform a quick comparison with it on/off in file-ton to see if there is any effect.

File: ewkcoffea/modules/selection_wwz.py (hunk 6 above)
Comment (MatthewDittrich) [reply ewkcoffea/modules/selection_wwz.py:263]: I have them. there are some differences

File: analysis/wwz/wwz4l.py (hunk 10 above)
Comment (MatthewDittrich) [reply analysis/wwz/wwz4l.py:388]: This is old comment I forgot to remove.

File: analysis/wwz/wwz4l.py (hunk 12 above)
Comment (MatthewDittrich) [reply analysis/wwz/wwz4l.py:862]: yes, but this is done on line 927 for when the year is not 2022

File: ewkcoffea/modules/objects_wwz.py (hunk 1 above)
Comment (kmohrman) [reply ewkcoffea/modules/objects_wwz.py]: Thanks for updating that. However, the if logic here is now pretty hard to follow (for me anyway). I think it would first of all be good to just pass your `is2022` variable to this function so that we don't have to check if "2022" is in the year. 

Then secondly I think you can move the final if into the else (that went with the first if). So instead of:
```if ("2022" in year): 
        mask_return = (mask & ele.mvaIso_WP80 & ele.convVeto)
    else: 
        mask_return = (mask & mask_run2)
    if tight and ("2022" not in year): 
        mask_return = (mask_return & ele.convVeto & (ele.tightCharge == get_ec_param("wwz_pres_e_tightCharge")))
```
It would be something like this:
```
    if  is2022: 
        mask_return = (mask & ele.mvaIso_WP80 & ele.convVeto)
    else: 
        if tight: 
            mask_return = (mask_return & ele.convVeto & (ele.tightCharge == get_ec_param("wwz_pres_e_tightCharge")))
        else:
            mask_return = (mask & mask_run2)
```

Comment (kmohrman) [reply ewkcoffea/modules/objects_wwz.py]: Though additionally, are you sure this does not change the behavior for R2? Where do you apply the mask_run2 in the case of tight?

Comment (kmohrman) [reply ewkcoffea/modules/objects_wwz.py]: In fact, at this point, I think it might be cleaner to just split this function into two different functions. There is really not much overlapping code here (just the first 4 lines only).

Comment (kmohrman) [reply ewkcoffea/modules/objects_wwz.py]: Actually, it seems we never use this function for R2 _without_ passing Tight=True. So let's just get rid of the tight option all together and always apply that for R2. That will make the function a lot cleaner, and I think we can leave it as one function. 

But please do pass the is2022 variable and restructure the if statements as described above.

File: ewkcoffea/modules/objects_wwz.py (hunk 3 above)
Comment (kmohrman) [reply ewkcoffea/modules/objects_wwz.py]: Passing the is2022 argument would be good here too.

File: ewkcoffea/modules/selection_wwz.py (hunk 4 above)
Comment (kmohrman) [reply ewkcoffea/modules/selection_wwz.py]: Was there any effect?

File: ewkcoffea/modules/selection_wwz.py (hunk 6 above)
Comment (kmohrman) [reply ewkcoffea/modules/selection_wwz.py]: Please pass the is2022 flag for this.

File: analysis/wwz/wwz4l.py [hunk 13]
```diff
@@ -169,17 +169,24 @@ def columns(self):
...
         # Dataset parameters
-        dataset = events.metadata["dataset"]
+        json_name = events.metadata["dataset"]
+
+        isData             = self._samples[json_name]["isData"]
+        histAxisName       = self._samples[json_name]["histAxisName"]
+        year               = self._samples[json_name]["year"]
+        xsec               = self._samples[json_name]["xsec"]
+        sow                = self._samples[json_name]["nSumOfWeights"]
+        if "2022" in year:
```
Comment (kmohrman) [inline analysis/wwz/wwz4l.py:179]: It is a much stronger check and less error prone to do `if year in ["2022", "2022EE"]` rather than `if "2022" in year`

**Comment by kmohrman** [comment]
The CI seems to be failing on the R3 step. Could you check into what's causing that?

File: analysis/wwz/wwz4l.py (hunk 12 above)
Comment (kmohrman) [reply analysis/wwz/wwz4l.py:862]: Ok. I will try to think about how to make this part of the code less confusing (I don't like how I'm checking which histograms make sense for which categories). Anyway, this is fine for now I think.

File: ewkcoffea/modules/selection_wwz.py (hunk 4 above)
Comment (MatthewDittrich) [reply ewkcoffea/modules/selection_wwz.py]: very negligible change (5 orders of magnitude difference)

Comment (kmohrman) [reply ewkcoffea/modules/selection_wwz.py]: For di lepton for 4l?

Comment (MatthewDittrich) [reply ewkcoffea/modules/selection_wwz.py]: dilepton

**Comment by kmohrman** [comment]
Are you sure the change is yields was due to the filter flag requirement? I just tried locally swapping in your new requirement, and got no change in yields for the ref file. 

Can you try running on the reference file with and without the new flags, and verify if that change in yields in the reference file is really due to the filter flag requirement?

File: ewkcoffea/modules/objects_wwz.py (hunk 1 above)
Comment (kmohrman) [reply ewkcoffea/modules/objects_wwz.py]: Thanks for the update. I think this looks a _lot_ easier to read now. One last request (sorry), could you remove the `year` argument now? Since it's no longer used by the function.

File: analysis/wwz/wwz4l.py (hunk 7 above)
Comment (kmohrman) [reply analysis/wwz/wwz4l.py]: It seems the dataset split thing is still happening outside the if block. I would put all of this stuff inside the if since it's only relevant for data. Also, one comment describing what it's doing would be useful.

File: ewkcoffea/modules/selection_wwz.py (hunk 4 above)
Comment (kmohrman) [reply ewkcoffea/modules/selection_wwz.py]: Ok, I'm good with removing it, as discussed on matter most.

**Comment by kmohrman** [comment]
Thank you to @MatthewDittrich for all of the work on this PR. For future reference, here is a summary of the main points:
* Run 3 can now be processed (by the main analysis processor) and associated yields and datacards scripts:
  * See the update reported at SMP-VV analysis meeting [here](https://indico.cern.ch/event/1402880/contributions/5897023/attachments/2847442/4978967/Dittrich_SMPApril30_04222024.pdf) for details (though note the VHnobb overlap removal was applied _after_ those slides were prepared)
  * See also [topcoffea PR#23](https://github.com/TopEFT/topcoffea/pull/23) for the corresponding updates for R2 corrections. 
  * R3 systematics are not yet applied. 
  * The data-driven background estimation is currently applied in the same way as R2. 
  * CI has been added to test a R2 root file. 
* Small updates to the R2 BDT (v5) SR binning have been applied (amounting to a >5% improvement in the significance) 
* Json files and corresponding configs for the HPG copies of the R2 skims have been added. 

For future reference, here is a summary of the relevant significances:
* 2022 (no systematics, no data-driven background estimation): 1.52453
* 2022 (no systematics, with data-driven background estimation): 1.47384
* R2 cut-based (with systematics, with data-driven background estimation): 3.45205
* R2 BDT (with systematics, with data-driven background estimation): 4.5185
